astropy/io/misc/tests/test_hdf5.py

```diff
@@ -401,17 +401,98 @@ def test_preserve_serialized(tmpdir):
+def test_preserve_serialized_compatibility_mode(tmpdir):
     test_file = str(tmpdir.join('test.hdf5'))
 
     t1 = Table()
-    t1['a'] = Column(data=[1, 2, 3])
-    t1.meta["meta"] = "0" * (2**16 + 1)
+    t1['a'] = Column(data=[1, 2, 3], unit="s")
+    t1['a'].meta['a0'] = "A0"
+    t1['a'].meta['a1'] = {"a1": [0, 1]}
+    t1['a'].format = '7.3f'
+    t1['a'].description = 'A column'
+    t1.meta['b'] = 1
+    t1.meta['c'] = {"c0": [0, 1]}
 
     with catch_warnings() as w:
-        t1.write(test_file, path='the_table', serialize_meta=True, overwrite=True)
-    assert len(w) == 1
+        t1.write(test_file, path='the_table', serialize_meta=True,
+                 overwrite=True, compatibility_mode=True)
+
     assert str(w[0].message).startswith(
+        "compatibility mode for writing is deprecated")
+
+    t2 = Table.read(test_file, path='the_table')
+
+    assert t1['a'].unit == t2['a'].unit
+    assert t1['a'].format == t2['a'].format
+    assert t1['a'].description == t2['a'].description
+    assert t1['a'].meta == t2['a'].meta
+    assert t1.meta == t2.meta
+
+
+@pytest.mark.skipif('not HAS_H5PY or not HAS_YAML')
+def test_preserve_serialized_in_complicated_path(tmpdir):
+    test_file = str(tmpdir.join('test.hdf5'))
+
+    t1 = Table()
+    t1['a'] = Column(data=[1, 2, 3], unit="s")
+    t1['a'].meta['a0'] = "A0"
+    t1['a'].meta['a1'] = {"a1": [0, 1]}
+    t1['a'].format = '7.3f'
+    t1['a'].description = 'A column'
+    t1.meta['b'] = 1
+    t1.meta['c'] = {"c0": [0, 1]}
+
+    t1.write(test_file, path='the_table/complicated/path', serialize_meta=True,
+             overwrite=True)
+
+    t2 = Table.read(test_file, path='the_table/complicated/path')
+
+    assert t1['a'].format == t2['a'].format
+    assert t1['a'].unit == t2['a'].unit
+    assert t1['a'].description == t2['a'].description
+    assert t1['a'].meta == t2['a'].meta
+    assert t1.meta == t2.meta
+
+@pytest.mark.skipif('not HAS_H5PY or not HAS_YAML')
+def test_metadata_very_large(tmpdir):
+    """Test that very large datasets work, now!"""
+    test_file = str(tmpdir.join('test.hdf5'))
+
+    t1 = Table()
+    t1['a'] = Column(data=[1, 2, 3], unit="s")
+    t1['a'].meta['a0'] = "A0"
+    t1['a'].meta['a1'] = {"a1": [0, 1]}
+    t1['a'].format = '7.3f'
+    t1['a'].description = 'A column'
+    t1.meta['b'] = 1
+    t1.meta['c'] = {"c0": [0, 1]}
+    t1.meta["meta_big"] = "0" * (2 ** 16 + 1)
+    t1.meta["meta_biggerstill"] = "0" * (2 ** 18)
+
+    t1.write(test_file, path='the_table', serialize_meta=True, overwrite=True)
+
+    t2 = Table.read(test_file, path='the_table')
+
+    assert t1['a'].unit == t2['a'].unit
+    assert t1['a'].format == t2['a'].format
+    assert t1['a'].description == t2['a'].description
+    assert t1['a'].meta == t2['a'].meta
+    assert t1.meta == t2.meta
+
+
+@pytest.mark.skipif('not HAS_H5PY or not HAS_YAML')
+def test_metadata_very_large_fails_compatibility_mode(tmpdir):
+    """Test that very large datasets work, now!"""
```

Comment:
For the record this docstring is misleading.  Not an absolutely required fix though.